ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/storage/BonsaiSnapshotWorldStateKeyValueStorage.java

```diff
@@ -0,0 +1,217 @@
+  }
+
+  @Override
+  public Optional<Bytes> getWorldStateRootHash() {
+    return isClosed.get() ? Optional.empty() : super.getWorldStateRootHash();
+  }
+
+  @Override
+  public Optional<Hash> getWorldStateBlockHash() {
+    return isClosed.get() ? Optional.empty() : super.getWorldStateBlockHash();
+  }
+
+  @Override
+  public Optional<Bytes> getStorageValueBySlotHash(final Hash accountHash, final Hash slotHash) {
+    return isClosed.get()
+        ? Optional.empty()
+        : super.getStorageValueBySlotHash(accountHash, slotHash);
+  }
+
+  @Override
+  public Optional<Bytes> getStorageValueBySlotHash(
+      final Supplier<Optional<Hash>> storageRootSupplier,
+      final Hash accountHash,
+      final Hash slotHash) {
+    return isClosed.get()
+        ? Optional.empty()
+        : super.getStorageValueBySlotHash(storageRootSupplier, accountHash, slotHash);
+  }
+
+  @Override
+  public boolean isWorldStateAvailable(final Bytes32 rootHash, final Hash blockHash) {
+    return !isClosed.get() && super.isWorldStateAvailable(rootHash, blockHash);
+  }
+
+  @Override
+  public void clear() {
+    // snapshot storage does not implement clear
+    throw new StorageException("Snapshot storage does not implement clear");
+  }
+
+  @Override
+  public void clearFlatDatabase() {
+    // snapshot storage does not implement clear
+    throw new StorageException("Snapshot storage does not implement clear");
+  }
+
+  @Override
+  public void clearTrieLog() {
+    // snapshot storage does not implement clear
+    throw new StorageException("Snapshot storage does not implement clear");
+  }
+
+  @Override
+  public void onCloseStorage() {
+    try {
+      // when the parent storage clears, close regardless of subscribers
+      doClose();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void onClearStorage() {
+    try {
+      // when the parent storage clears, close regardless of subscribers
+      doClose();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void onClearFlatDatabaseStorage() {
+    // when the parent storage clears, close regardless of subscribers
+    try {
+      doClose();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void onClearTrieLog() {
+    // when the parent storage clears, close regardless of subscribers
+    try {
+      doClose();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected synchronized void doClose() throws Exception {
+    if (!isClosed.get()) {
+      // alert any subscribers we are closing:
+      subscribers.forEach(BonsaiStorageSubscriber::onCloseStorage);
+
+      // close all of the SnappedKeyValueStorages:
+      accountStorage.close();
+      codeStorage.close();
+      storageStorage.close();
+      trieBranchStorage.close();
+
```

Comment:
We are missing closing trieLogStorage here.  snapshots did not close trieLogStorage

Reply:
Trielog is never a copy but it is always the real column so if we close we will close the real column. On the other hand we forgot it here  https://github.com/hyperledger/besu/pull/5123/files/f96ca374c433439980b64e3383901daffb1e98fe#diff-fa9b23a353ef5038c992c17a8cfe5e6210cc41369dada665044bc7633e136ab0R457

Reply:
sorry - meant for this comment to be on BonsaiWorldStateKeyValueStorage.doClose() not the snapshot doClose()